proposals/2020-07-28_keyring-cmm/proposal.md

```diff
@@ -0,0 +1,209 @@
+we expect to update the Default CMM to a configuration which composes such CMMs together provide
+useful properties.
+Any update to the composition defined by Default CMM SHOULD be backwards compatable.
+
+To ensure that its use is correct by construction,
+the instantiation of the Default CMM MUST NOT take any options
+except for the keyring that should be composed with the Default CMM.
+
+### Keyring CMM
+
+The Keyring CMM is a base CMM implementation that
+[gets encryption materials](../../framework/cmm-interface.md#get-encryption-materials)
+and [decrypts materials](../../framework/cmm-interface.md#decrypt-materials)
+through use of a [keyring](../../framework/keyring-interface.md) configured on instantiation.
+
+Any desired CMM implementation that intends to use keyrings to ensure valid materials SHOULD
+compose with the Keyring CMM.
+
+### Master Key Provider CMM
+
+Any implementation that previously supported
+[master key providers](../../framework/master-key-provider-interface.md)
+MUST have a Master Key Provider CMM.
+Any implementation that does not support master key providers MUST NOT
+provide a Master Key Provider CMM.
+
+This CMM SHOULD NOT be used except for backwards compatability reasons.
+
+The Master Key Provider CMM is a base CMM implementation that
+[gets encryption materials](../../framework/cmm-interface.md#get-encryption-materials)
+and [decrypts materials](../../framework/cmm-interface.md#decrypt-materials)
+through use of a [master key provider](../../framework/master-key-provider-interface.md)
+configured on instantiation.
+
+Any desired CMM implementation that intends to use master key providers to ensure valid materials
+SHOULD compose with the Master Key Provider CMM.
+
+## Reference-level Explanation
+
+### Default CMM
+
+On initialization, the caller MUST provide exactly one of the following:
+
+- [Keyring](../../framework/keyring-interface.md)
+- If this AWS Encryption SDK implementations provides a [Master Key Provider CMM](#master-key-provider-cmm),
+  a [Master Key Provider](../../framework/master-key-provider-interface.md)
+
+The Default CMM MUST NOT accept any additional configuration.
+
+It MUST construct a CMM in the following manner:
+
+- If a keyring was supplied, initialize a [Keyring CMM](#keyring-cmm) with the provided keyring.
+- If a master key provider was supplied, initialize a [Master Key Provider CMM](#keyring-cmm)
+  with the provided master key provider.
+
+This CMM MUST NOT offer any additional features beyond the composed CMM created
+above.
+
+Functionality exposed by the AWS Encryption SDK MUST NOT require the Default CMM.
+
+### Keyring CMM
+
+The specification of the Keyring CMM MUST be the specification of the
+Default CMM prior to this change.
+
+### Master Key Provider CMM
+
+The specification of the Master Key Provider CMM MUST be the specification of
+the Default CMM, except describing use of a master key provider instead of a keyring
```

Comment:
```suggestion
The specification of the Master Key Provider CMM MUST be the specification of the Default CMM,
except describing the use of a master key provider instead of a keyring
```